src/Microsoft.DotNet.Build.Tasks/DownloadFilesFromUrl.cs

```diff
@@ -0,0 +1,144 @@
+using Microsoft.Build.Framework;
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Net;
+using System.Collections.Generic;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public sealed class DownloadFilesFromUrl : BuildTask
+    {
+        private static readonly HttpClient s_client = new HttpClient(GetHttpHandler());
+
+        /// <summary>
+        /// The items to download.
+        /// Url and DestinationFile are required in the item's metadata, DestinationDir is optional.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Items { get; set; }
+
+        /// <summary>
+        /// The destination directory for all the items to be downloaded.
+        /// This parameter can be overriden per item, if an item has DestinationDir in it's metadata, we will take that directory.
+        /// If both, the item's and this property are empty, the items will be downloaded to the working directory.
+        /// </summary>
+        public string DestinationDir { get; set; }
+
+        /// <summary>
+        /// The default is to fail on error, set this to true if you want to just warn.
+        /// </summary>
+        public bool TreatErrorsAsWarnings { get; set; }
+
+        /// <summary>
+        /// The list of files created. It is not guaranted that all the input items will be successfully downloaded
+        /// when TreatErrorsAsWarings is set to true.
+        /// </summary>
+        [Output]
+        public ITaskItem[] FilesCreated { get; set; }
+
+        public override bool Execute()
+        {
+            var filesCreated = new List<ITaskItem>();
+            foreach (var item in Items)
+            {
+                string downloadSource = item.GetMetadata("Url");
+                if (string.IsNullOrWhiteSpace(downloadSource))
+                {
+                    if (TreatErrorsAsWarnings)
+                    {
+                        Log.LogWarning($"Item {item.ItemSpec} is missing Url to download from in it's metadata.");
+                        continue;
+                    }
+                    else
+                    {
+                        Log.LogError($"Item {item.ItemSpec} is missing Url to download from in it's metadata.");
+                        return false;
+                    }
+                }
+
+                string fileName = item.GetMetadata("DestinationFile");
```

Comment:
don't we need to set this if it isn't found? Wouldn't we default to the filename of the requested file.

Reply:
But we actually don't have the filename of the requested file, as it is a url and the url doesn't always in the filename that we want to download i.e to download a symbol package it is `https://feedurl/symbols/packageId/packageVersion` so we don't really have a way to know the default filename.

Reply:
There will be some cases were the Url will have the filename at the end, but that is not always the case, that is why I wanted to force people to give the destination file name (which they should always know).

Reply:
What do you download in that case?

Reply:
Whatever the HTTP GET request returns, which in the case of the symbol server it returns a symbols.nupkg file. If the request doesn't return anything or throws an error then we would just output a a warning or an error, depending on the `TreatErrorsAsWarnings` parameter.

Reply:
OK perhaps default the filename based on the url if it has a filename.

Reply:
OK if we can't find a filename in the URL and the DestinationFile metadata is empty then we throw the error.

Reply:
Yeah otherwise we have an empty filename.

Reply:
But empty file name will cause an exception when trying to create the file, do we want to handle that error ourselves or should we leave it to `File.OpenWrite` ?

Reply:
I think we should handle it ourselves so we can give a better error message.

Reply:
OK.